src/tasks/EmitBundleTask/EmitBundleBase.cs

```diff
@@ -0,0 +1,381 @@
+            addPreallocatedResources.AppendLine($"    mono_bundled_resources_add (bundledDataResources, {dataCount});");
+        }
+
+        outputUtf8Writer.Write(Utils.GetEmbeddedResource("mono-bundled-resource-preallocation-and-registration.template")
+                                .Replace("%PreallocatedStructs%", preallocatedSource.ToString())
+                                .Replace("%PreallocatedResources%", preallocatedResources.ToString())
+                                .Replace("%BundleRegistrationFunctionName%", bundleRegistrationFunctionName)
+                                .Replace("%AddPreallocatedResources%", addPreallocatedResources.ToString()));
+    }
+
+    private static void BundleFileToCSource(string symbolName, FileStream inputStream, Stream outputStream)
+    {
+        // Emits a C source file in the same format as "xxd --include". Example:
+        //
+        // unsigned char Some_File_dll[] = {
+        //   0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x0a
+        // };
+        // unsigned int Some_File_dll_len = 6;
+
+        var buf = new byte[4096];
+        int bytesRead;
+        var generatedArrayLength = 0;
+        var bytesEmitted = 0;
+
+        using var outputUtf8Writer = new StreamWriter(outputStream, Utf8NoBom);
+
+        outputUtf8Writer.Write($"unsigned char {symbolName}_data[] = {{");
+        outputUtf8Writer.Flush();
+        while ((bytesRead = inputStream.Read(buf, 0, buf.Length)) > 0)
+        {
+            for (var i = 0; i < bytesRead; i++)
+            {
+                if (bytesEmitted++ % 12 == 0)
+                {
+                    outputStream.Write(NewLineAndIndentation, 0, NewLineAndIndentation.Length);
+                }
+
+                var byteValue = buf[i];
+                outputStream.Write(HexToUtf8Lookup, byteValue * 6, 6);
+            }
+
+            generatedArrayLength += bytesRead;
+        }
+
+        outputUtf8Writer.WriteLine("0\n};");
+        outputUtf8Writer.WriteLine($"unsigned int {symbolName}_len = {generatedArrayLength};");
+        outputUtf8Writer.Flush();
+        outputStream.Flush();
+
+        symbolDataLen.Add(symbolName, generatedArrayLength);
+    }
+
+    private static string ToSafeSymbolName(string destinationFileName)
+    {
+        // Since destinationFileName includes a content hash, we can safely strip off the directory name
+        // as the filename is always unique enough. This avoid disclosing information about the build
+        // file structure in the resulting symbols.
+        var filename = Path.GetFileName(destinationFileName);
+
+        // Equivalent to the logic from "xxd --include"
+        var sb = new StringBuilder();
+        foreach (var c in filename)
+        {
+            sb.Append(IsAlphanumeric(c) ? c : '_');
+        }
+
+        return sb.ToString();
+    }
+
+    private static string GetFileType(string destinationFileName)
+    {
+        if (destinationFileName.EndsWith(".resources.dll"))
```

Comment:
It seems like some of the satellite assembly resources may not always end in `.resources.dll`, should figure out what extensions we should support, or more importantly, opting to determine file type based on metadata attached to the file before the file reaches this task.